text/0000-manifest-list-buildpacks.md

```diff
@@ -0,0 +1,133 @@
+
+Here is an example of what the dependencies would look like.
+
+```toml
+  [[metadata.dependencies]]
+    cpe = "cpe:2.3:a:golang:go:1.18.4:*:*:*:*:*:*:*"
+    id = "go"
+    licenses = ["BSD-3-Clause"]
+    name = "Go"
+    stacks = ["*"]
+    strip-components = 1
+    version = "1.18.4"
+    os = "linux"
+    architecture = "amd64"
+    purl = "pkg:generic/go@go1.19?checksum=c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5&download_url=https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+    sha256 = "c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5"
+    source = "https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+    source_sha256 = "c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5"
+    uri = "https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+
+
+  [[metadata.dependencies]]
+    cpe = "cpe:2.3:a:golang:go:1.18.4:*:*:*:*:*:*:*"
+    id = "go"
+    licenses = ["BSD-3-Clause"]
+    name = "Go"
+    stacks = ["*"]
+    strip-components = 1
+    version = "1.18.4"
+    os = "linux"
+    architecture = "arm64"
+    purl = "pkg:generic/go@go1.19?checksum=35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f&download_url=https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+    sha256 = "35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f"
+    source = "https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+    source_sha256 = "35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f"
+    uri = "https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+```
+
+### Github actions and workflows
+
+I haven't looked at github workflows for all buildpacks, but I believe [jam](https://github.com/paketo-buildpacks/jam) is being used to package buildpacks. I did see skopeo being used to push images ([cpython](https://github.com/paketo-buildpacks/cpython/blob/main/.github/workflows/push-buildpackage.yml#L62)).
+
+My proposal is to update jam to create buildpacks for amd64 and arm64. I haven't used jam yet, so I don't have specific code changes for it and would appreciate thoughts on how this could be added.
+
+These can then be pushed using skopeo with the above mentioned architecture-specific tags. Finally some `docker manifest` commands can be used to create the manifest list image and then push it to the registry.
+
+Users would have the same experience using the manifest list images on amd64, while allowing those who want/need arm64 images to benefit as well.
+
+
+## Prior Art
+
+As mentioned above @dmikusa put together [this](https://github.com/dmikusa/paketo-arm64) excellent guide for building arm64 images with paketo buildpacks. This repo, based on the aforementioned example, is publishing arm version of paketo builders and buildpacks. While it is great to see community involvement in pushing this forward, I think it is time for the paketo maintainers to agree on a path forward so the community can start contributing towards that goal.
+
+
+## Unresolved Questions and Bikeshedding
```

Comment:
The biggest issue I see with arm support on buildpacks is compiling dependencies (like cpython) for arm64.

We can compile the buildpacks themselves via Go's built-in cross-compilation, but that doesn't extend to arbitrary compilation (e.g. Cpython is written in C and uses `gcc` to compile).

Many language families use prebuilt, upstream dependencies (e.g. Go), so if they provide pre-built versions for the arm64 architecture we can just use them using the changes you describe above.

Some buildpacks (like cpython) support building from source during build-time, so that would be a workaround too. But for buildpacks which use pre-compiled dependencies and don't have the escape hatch of compiling during build (e.g. web-servers, ruby), they just won't work on arm64 until we have the ability to compile the dependencies for arm64.

Reply:
As a project, we have access to ARM runners so we can build packages where necessary. That will probably require work, not sure if the existing build will just run on Arm64, there could be differences.

I believe that we start with what Buildpacks we have though and work up from there. The Arm64 support will have to be beta for a while well we test it and get all the Buildpacks functional on it.

Reply:
There's a thread below which has more context on exactly what we compile/test on which runners. I'll close this thread in favor of that.